R/summarize.R

```diff
@@ -1,29 +1,85 @@
 #' @export
-#' @importFrom dplyr summarise_
+#' @importFrom dplyr bind_cols summarise_ select
+#' @importFrom purrr map_chr map_df
 #' @importFrom crunch crtabs
 summarise_.CrunchDataset <- function (.data, ..., .dots) {
     dots <- all_dots(.dots, ..., all_named = TRUE)
-    fmla <- dots_to_formula(dots, groups(.data))
-    out <- as_tibble(crtabs(fmla, data=.data))
-    ## Sort the return columns based on the request order
-    out <- out[, c(setdiff(names(out), names(dots)), names(dots))]
+    unweighted <- dots %>% map_chr(~as.character(.$expr)[[1]]) == "unweighted_n"
+    counts <- dots[unweighted]
+    non_counts <- dots[!unweighted]
+    fmla <- dots_to_formula(non_counts, groups(.data))
+    
+    # When there are no groups or summary functions, we can't 
+    # use crtabs to get a cube, but unweighted_n() is equivalent to nrow(ds)
+    # Otherwise we can use crtabs to get the cube and extract the unweighted counts
+    # from the cube. This doesn't work for weighted counts because they 
+    # are not included in some cubes. So in that case we rely on the the server
+    # `n` function. 
+    if (length(non_counts) == 0 && length(groups(.data)) == 0) {
+        out <- map_df(counts, ~nrow(.data))
```

Comment:
1) This is a weird edge case. I'd move the large comment block inside the `if`, and then start the `else` below with `# The usual case`, just to make that clear.

2) Technically you could construct a cube query to get this, but `nrow()` is much more intuitive.

3) Explain here why you're doing map_df, for the pathological case where not only does the user request `unweighted_n()` and nothing else but they also request it multiple times! Then you don't have to explain it below.